[â¬…ï¸ Back to Table of Contents](../../../../index.md)

# ğŸ“„ `no-invalid-this.ts`

## ğŸ“Š Analysis Summary

| Metric | Count |
|--------|-------|
| ğŸ”§ Functions | 0 |
| ğŸ§± Classes | 0 |
| ğŸ“¦ Imports | 6 |
| ğŸ“Š Variables & Constants | 3 |
| âœ¨ Decorators | 0 |
| ğŸ”„ Re-exports | 0 |
| âš¡ Async/Await Patterns | 0 |
| ğŸ’  JSX Elements | 0 |
| ğŸŸ¢ Vue Composition API | 0 |
| ğŸ“ Interfaces | 0 |
| ğŸ“‘ Type Aliases | 2 |
| ğŸ¯ Enums | 0 |

## ğŸ“š Table of Contents

- [Imports](#imports)
- [Variables & Constants](#variables-constants)
- [Type Aliases](#type-aliases)

## ğŸ› ï¸ File Location:
ğŸ“‚ **`packages/eslint-plugin/src/rules/no-invalid-this.ts`**

## ğŸ“¦ Imports

| Name | Source |
|------|--------|
| `TSESTree` | `@typescript-eslint/utils` |
| `AST_NODE_TYPES` | `@typescript-eslint/utils` |
| `InferMessageIdsTypeFromRule` | `../util` |
| `InferOptionsTypeFromRule` | `../util` |
| `createRule` | `../util` |
| `getESLintCoreRule` | `../util/getESLintCoreRule` |


---

## Variables & Constants

| Name | Type | Kind | Value | Exported |
|------|------|------|-------|----------|
| `defaultOptions` | `Options` | const | `[{ capIsConstructor: true }]` | âœ— |
| `thisIsValidStack` | `boolean[]` | const | `[]` | âœ— |
| `thisIsValidHere` | `boolean` | const | `thisIsValidStack[thisIsValidStack.length - 1]` | âœ— |


---


---

## Type Aliases

### `Options`

```ts
type Options = InferOptionsTypeFromRule<typeof baseRule>;
```

### `MessageIds`

```ts
type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
```


---