[â¬…ï¸ Back to Table of Contents](../../../../index.md)

# ðŸ“„ `no-invalid-this.ts`

## ðŸ“Š Analysis Summary

| Metric | Count |
|--------|-------|
| ðŸ”§ Functions | 0 |
| ðŸ§± Classes | 0 |
| ðŸ“¦ Imports | 6 |
| ðŸ“Š Variables & Constants | 3 |
| âœ¨ Decorators | 0 |
| ðŸ”„ Re-exports | 0 |
| âš¡ Async/Await Patterns | 0 |
| ðŸ’  JSX Elements | 0 |
| ðŸŸ¢ Vue Composition API | 0 |
| ðŸ“ Interfaces | 0 |
| ðŸ“‘ Type Aliases | 2 |
| ðŸŽ¯ Enums | 0 |

## ðŸ“š Table of Contents

- [Imports](#imports)
- [Variables & Constants](#variables-constants)
- [Type Aliases](#type-aliases)

## ðŸ› ï¸ File Location:
ðŸ“‚ **`packages/eslint-plugin/src/rules/no-invalid-this.ts`**

## ðŸ“¦ Imports

| Name | Source |
|------|--------|
| `TSESTree` | `@typescript-eslint/utils` |
| `AST_NODE_TYPES` | `@typescript-eslint/utils` |
| `InferMessageIdsTypeFromRule` | `../util` |
| `InferOptionsTypeFromRule` | `../util` |
| `createRule` | `../util` |
| `getESLintCoreRule` | `../util/getESLintCoreRule` |


---

## Variables & Constants

| Name | Type | Kind | Value | Exported |
|------|------|------|-------|----------|
| `defaultOptions` | `Options` | const | `[{ capIsConstructor: true }]` | âœ— |
| `thisIsValidStack` | `boolean[]` | const | `[]` | âœ— |
| `thisIsValidHere` | `boolean` | const | `thisIsValidStack[thisIsValidStack.length - 1]` | âœ— |


---

## ðŸ”§ Functions

> No functions found in this file.


---

## Type Aliases

### `Options`

```ts
type Options = InferOptionsTypeFromRule<typeof baseRule>;
```

### `MessageIds`

```ts
type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
```


---