[â¬…ï¸ Back to Table of Contents](../../../../index.md)

# ðŸ“„ `no-invalid-this.ts`

## ðŸ“š Table of Contents

- [Imports](#imports)
- [Type Aliases](#type-aliases)

## ðŸ“Š Analysis Summary

- **Functions**: 0
- **Classes**: 0
- **Imports**: 6
- **Interfaces**: 0
- **Type Aliases**: 2

## ðŸ› ï¸ File Location:
ðŸ“‚ **`packages/eslint-plugin/src/rules/no-invalid-this.ts`**

## ðŸ“¦ Imports

| Name | Source |
|------|--------|
| `TSESTree` | `@typescript-eslint/utils` |
| `AST_NODE_TYPES` | `@typescript-eslint/utils` |
| `InferMessageIdsTypeFromRule` | `../util` |
| `InferOptionsTypeFromRule` | `../util` |
| `createRule` | `../util` |
| `getESLintCoreRule` | `../util/getESLintCoreRule` |


---

## ðŸ”§ Functions

> No functions found in this file.


---

## Classes

> No classes found in this file.


---

## Interfaces

> No interfaces found in this file.


---

## Type Aliases

### `Options`

```ts
type Options = InferOptionsTypeFromRule<typeof baseRule>;
```

### `MessageIds`

```ts
type MessageIds = InferMessageIdsTypeFromRule<typeof baseRule>;
```


---